File: src/ISADotnet/API/Update.fs
namespace ISADotNet.API

open ISADotNet

module Option =
 
    /// If the value matches the default, a None is returned, else a Some is returned
    let fromValueWithDefault d v =
        if d = v then None
        else Some v

    /// Applies the function f on the value of the option if it exists, else applies it on the default value. If the result value matches the default, a None is returned
    let mapDefault (d : 'T) (f: 'T -> 'T) (o : 'T option) =
        match o with
        | Some v -> f v
        | None   -> f d
        |> fromValueWithDefault d

module Update =

    open System
    open Microsoft.FSharp.Reflection

    /// matches if the matched object can be parsed to Some 'a and returns it.
    let private (|SomeObj|_|) =
        /// create generalized option type
        let ty = typedefof<option<_>>
        fun (a:obj) ->
            /// Check for nulls otherwise 'a.GetType()' would fail
            if isNull a 
            then 
                None 
            else
                let aty = a.GetType()
                /// Get option'.Value
                let v = aty.GetProperty("Value")
                if aty.IsGenericType && aty.GetGenericTypeDefinition() = ty then
                    /// return value if existing
                    Some(v.GetValue(a, [| |]))
                else 
                    None
    
    let private makeOptionValue typey v isSome =
        let optionType = typeof<unit option>.GetGenericTypeDefinition().MakeGenericType([|typey|])
        let cases = FSharp.Reflection.FSharpType.GetUnionCases(optionType)
        let cases = cases |> Array.partition (fun x -> x.Name = "Some")
        let someCase = fst cases |> Array.exactlyOne
        let noneCase = snd cases |> Array.exactlyOne
        let relevantCase, args =
            match isSome with
            | true -> someCase, [| v |]
            | false -> noneCase, [| |]
        FSharp.Reflection.FSharpValue.MakeUnion(relevantCase, args)

    /// updates oldRT with newRT by replacing all values, but appending all lists.
    ///
    /// newRTList@oldRTList
    let rec private updateAppend (oldVal: obj) (newVal:obj) = 
        
        /// match all field Values and try to cast them to types.
        match oldVal with
        /// Strings are IEnumerable Chars but should not be appenden. So these have to be handled first.
        | :? String ->
            newVal
        | SomeObj(oldInternal) ->
            match oldInternal with
            | :? String -> 
                newVal
            | :? System.Collections.IEnumerable ->
                let newOpt = newVal
                match newOpt with
                | SomeObj newInternal ->
                    updateAppend oldInternal newInternal
                    |> fun v -> makeOptionValue (oldInternal.GetType()) v true
                | _ -> oldVal
            | _ -> 
                newVal
        /// Match all IEnumarables, like list, array, seq. These should be appended.
        | :? System.Collections.IEnumerable -> 
            let oldSeq = oldVal
            let newSeq = newVal
            let fieldT = oldVal.GetType()
            /// Maps are IEnumerables but are not easily to append. TODO(?)
            let isMap =
                let genericMap = typeof<Map<_,_>>
                fieldT.Name = genericMap.Name
            /// t is the type of the IEnumerable elements.
            let t =
                /// element types are accessed differently for (list, seq) and Array. 
                if fieldT.IsArray then fieldT.GetElementType() else
                    oldVal.GetType().GetGenericArguments() |> Array.head
            /// This function accesses the append method of the list/array module and applies it accordingly to the element type.
            let appendGenericListsByType l1 l2 (t:Type) =
                // https://stackoverflow.com/questions/41253131/how-to-create-an-empty-list-of-a-specific-runtime-type
                System.Reflection.Assembly
                    .GetAssembly(typeof<_ list>)
                    .GetType(if fieldT.IsArray then "Microsoft.FSharp.Collections.ArrayModule" else "Microsoft.FSharp.Collections.ListModule")
                    .GetMethod("Append")
                    .MakeGenericMethod(t)
                    .Invoke(null, [|l1;l2|])
            /// If the IEnumerable is a map then we just replace with the new entry.
            if isMap then 
                newVal
            else
                let r = 
                    appendGenericListsByType oldSeq newSeq t
                r |> box
        /// All others do not need to be appended and can be replaced.
        | others -> 
            newVal


    /// updates oldRT with newRT by replacing all values, but only if the new value is not empty.
    let rec private updateOnlyByExisting (oldVal: obj) (newVal:obj) =      
        
        /// try to cast values to types to check for isEmpty according to type.
        match oldVal with 
        /// Check if newValue isNull = isEmpty
        | _ when newVal = null ->
            oldVal
        /// Handle OptionTypes
        // https://stackoverflow.com/questions/6289761/how-to-downcast-from-obj-to-optionobj
        /// Check of value is option, then check if new value isNone = isEmpty
        | SomeObj(oldInternal) ->
            let newOpt = newVal
            match newOpt with
            | SomeObj newInternal ->
                updateOnlyByExisting oldInternal newInternal
                |> fun v -> makeOptionValue (oldInternal.GetType()) v true
            | _ -> 
                oldVal
        /// Check if value is string, then check if new value is "" = isEmpty
        | :? String ->
            let newStr = newVal
            if string newStr = "" then oldVal else newStr
        /// https://stackoverflow.com/questions/47280544/determine-if-any-kind-of-list-sequence-array-or-ienumerable-is-empty
        /// Check if value is IEnumarable, then cast newValue to Seq and check if isEmpty
        | :? System.Collections.IEnumerable -> 
            let newSeq = newVal
            if newSeq :?> System.Collections.IEnumerable |> Seq.cast |> Seq.isEmpty
            then oldVal 
            else newSeq
        /// Others don't need to be checked as they have no clearly enough defined "empty" state
        | _ ->
            newVal
        
        

    /// This type specifies the exact manner on how complex types will be updated
    type UpdateOptions = 
        /// Updates all existing fields by replacing them with the corresponding new fields
        | UpdateAll
        /// Updates all existing fields by replacing them with the corresponding new fields if the new field is not empty.
        ///
        /// Empty string = ""; Empty IEnumerable (Set, List, Array, Seq, Map); Empty Option = None; null
        | UpdateByExisting
        /// Updates all existing fields by replacing them with the corresponding new fields, except any lists which will be appended with the new list shown first
        ///
        /// Maps are currently not appended
        | UpdateAllAppendLists

        /// This function will update recordType_1 with the values given in recordType_2 as specified by UpdateOption.
        member this.updateRecordType (recordType_1:'a) (recordType_2:'a) =    
            match this with
            | UpdateAll ->
                recordType_2
            | UpdateAllAppendLists ->
                (FSharp.Reflection.FSharpValue.GetRecordFields recordType_1,FSharp.Reflection.FSharpValue.GetRecordFields recordType_2)
                ||> Array.map2 updateAppend
                |> fun fields -> FSharpValue.MakeRecord(typeof<'a>, fields) :?> 'a
            | UpdateByExisting ->
                (FSharp.Reflection.FSharpValue.GetRecordFields recordType_1,FSharp.Reflection.FSharpValue.GetRecordFields recordType_2)
                ||> Array.map2 updateOnlyByExisting
                |> fun fields -> FSharpValue.MakeRecord(typeof<'a>, fields) :?> 'a
                

File: ts/ISA/ISA/Update.ts
import { Dictionary } from "../../fable_modules/fable-library-ts/MutableMap.js";
import { isIterable, defaultOf, IMap, structuralHash, equals } from "../../fable_modules/fable-library-ts/Util.js";
import { int32 } from "../../fable_modules/fable-library-ts/Int32.js";
import { isEmpty, iterate } from "../../fable_modules/fable-library-ts/Seq.js";
import { tryGetValue, addToDict } from "../../fable_modules/fable-library-ts/MapUtil.js";
import { Union, toString, FSharpRef } from "../../fable_modules/fable-library-ts/Types.js";
import { value as value_1, Option, some } from "../../fable_modules/fable-library-ts/Option.js";
import { append_generic, distinct_generic, isList_generic, isMap_generic } from "./Fable.js";
import { union_type, TypeInfo, makeUnion, option_type, makeGenericType, getUnionCases, name, obj_type } from "../../fable_modules/fable-library-ts/Reflection.js";
import { exactlyOne, partition } from "../../fable_modules/fable-library-ts/Array.js";

export function Dict_ofSeq<Key, T>(s: Iterable<[Key, T]>): IMap<Key, T> {
    const dict: IMap<Key, T> = new Dictionary<Key, T>([], {
        Equals: equals,
        GetHashCode: structuralHash,
    });
    iterate<[Key, T]>((tupledArg: [Key, T]): void => {
        addToDict(dict, tupledArg[0], tupledArg[1]);
    }, s);
    return dict;
}

export function Dict_tryFind<Key, T>(key: Key, dict: IMap<Key, T>): Option<T> {
    let patternInput: [boolean, T];
    let outArg: T = defaultOf();
    patternInput = ([tryGetValue(dict, key, new FSharpRef<T>((): T => outArg, (v: T): void => {
        outArg = v;
    })), outArg] as [boolean, T]);
    if (patternInput[0]) {
        return some(patternInput[1]);
    }
    else {
        return void 0;
    }
}

export function Dict_ofSeqWithMerge<T, Key>(merge: ((arg0: T, arg1: T) => T), s: Iterable<[Key, T]>): IMap<Key, T> {
    const dict: IMap<Key, T> = new Dictionary<Key, T>([], {
        Equals: equals,
        GetHashCode: structuralHash,
    });
    iterate<[Key, T]>((tupledArg: [Key, T]): void => {
        const k: Key = tupledArg[0];
        const v: T = tupledArg[1];
        const matchValue: Option<T> = Dict_tryFind<Key, T>(k, dict);
        if (matchValue == null) {
            addToDict(dict, k, v);
        }
        else {
            const v$0027: T = value_1(matchValue);
            dict.delete(k);
            addToDict(dict, k, merge(v$0027, v));
        }
    }, s);
    return dict;
}

/**
 * Get the type of the IEnumerable elements. E.g. for Array<'T> it would be 'T
 */
export function Update_isMapType(v: any): boolean {
    return isMap_generic<any>(v);
}

export function Update_isListType(v: any): boolean {
    return isList_generic<any>(v);
}

/**
 * Get the type of the IEnumerable elements. E.g. for Array<'T> it would be 'T
 */
export function Update_enumGetInnerType(v: any): any {
    return obj_type;
}

/**
 * updates oldRT with newRT by replacing all values, but appending all lists.
 * 
 * newRTList@oldRTList
 */
export function Update_updateAppend(oldVal: any, newVal: any): any {
    let matchResult: int32, oldInternal: any, others: any;
    if (typeof oldVal === "string") {
        matchResult = 0;
    }
    else if (isIterable(oldVal)) {
        let activePatternResult: Option<any>;
        const a: any = oldVal;
        activePatternResult = ((a == null) ? void 0 : (undefined));
        if (activePatternResult != null) {
            matchResult = 1;
            oldInternal = value_1(activePatternResult);
        }
        else {
            matchResult = 2;
        }
    }
    else {
        let activePatternResult_1: Option<any>;
        const a_1: any = oldVal;
        activePatternResult_1 = ((a_1 == null) ? void 0 : (undefined));
        if (activePatternResult_1 != null) {
            matchResult = 1;
            oldInternal = value_1(activePatternResult_1);
        }
        else {
            matchResult = 3;
            others = oldVal;
        }
    }
    switch (matchResult) {
        case 0:
            return newVal;
        case 1:
            if (typeof oldInternal! === "string") {
                return newVal;
            }
            else if (isIterable(oldInternal!)) {
                let activePatternResult_2: Option<any>;
                const a_2: any = newVal;
                activePatternResult_2 = ((a_2 == null) ? void 0 : (undefined));
                if (activePatternResult_2 != null) {
                    const newInternal: any = value_1(activePatternResult_2);
                    const v: any = Update_updateAppend(oldInternal!, newInternal);
                    const cases_1: [any[], any[]] = partition<any>((x: any): boolean => (name(x) === "Some"), getUnionCases(makeGenericType(option_type(obj_type), [obj_type])));
                    const patternInput = [exactlyOne<any>(cases_1[0]), [v]] as [any, any[]];
                    return makeUnion(patternInput[0], patternInput[1]);
                }
                else {
                    return oldVal;
                }
            }
            else {
                return newVal;
            }
        case 2: {
            const oldSeq = oldVal as Iterable<any>;
            const newSeq = newVal as Iterable<any>;
            const innerType: any = Update_enumGetInnerType(oldVal);
            if (Update_isMapType(oldVal)) {
                return newVal;
            }
            else {
                return distinct_generic<Iterable<any>, any>(append_generic<Iterable<any>, Iterable<any>>(oldSeq, newSeq));
            }
        }
        default:
            return newVal;
    }
}

/**
 * updates oldRT with newRT by replacing all values, but only if the new value is not empty.
 */
export function Update_updateOnlyByExisting(oldVal: any, newVal: any): any {
    if (equals(newVal, defaultOf())) {
        return oldVal;
    }
    else {
        let activePatternResult: Option<any>;
        const a: any = oldVal;
        activePatternResult = ((a == null) ? void 0 : (undefined));
        if (activePatternResult != null) {
            const oldInternal: any = value_1(activePatternResult);
            let activePatternResult_1: Option<any>;
            const a_1: any = newVal;
            activePatternResult_1 = ((a_1 == null) ? void 0 : (undefined));
            if (activePatternResult_1 != null) {
                const newInternal: any = value_1(activePatternResult_1);
                const v: any = Update_updateOnlyByExisting(oldInternal, newInternal);
                const cases_1: [any[], any[]] = partition<any>((x: any): boolean => (name(x) === "Some"), getUnionCases(makeGenericType(option_type(obj_type), [obj_type])));
                const patternInput = [exactlyOne<any>(cases_1[0]), [v]] as [any, any[]];
                return makeUnion(patternInput[0], patternInput[1]);
            }
            else {
                return oldVal;
            }
        }
        else if (typeof oldVal === "string") {
            const newStr: any = newVal;
            if (toString(newStr) === "") {
                return oldVal;
            }
            else {
                return newStr;
            }
        }
        else if (isIterable(oldVal)) {
            const newSeq: any = newVal;
            if (isEmpty<any>(newSeq as Iterable<any> as Iterable<any>)) {
                return oldVal;
            }
            else {
                return newSeq;
            }
        }
        else {
            return newVal;
        }
    }
}

/**
 * updates oldRT with newRT by replacing all values, but only if the new value is not empty.
 */
export function Update_updateOnlyByExistingAppend(oldVal: any, newVal: any): any {
    if (equals(newVal, defaultOf())) {
        return oldVal;
    }
    else {
        let activePatternResult: Option<any>;
        const a: any = oldVal;
        activePatternResult = ((a == null) ? void 0 : (undefined));
        if (activePatternResult != null) {
            const oldInternal: any = value_1(activePatternResult);
            let activePatternResult_1: Option<any>;
            const a_1: any = newVal;
            activePatternResult_1 = ((a_1 == null) ? void 0 : (undefined));
            if (activePatternResult_1 != null) {
                const newInternal: any = value_1(activePatternResult_1);
                const v: any = Update_updateOnlyByExistingAppend(oldInternal, newInternal);
                const cases_1: [any[], any[]] = partition<any>((x: any): boolean => (name(x) === "Some"), getUnionCases(makeGenericType(option_type(obj_type), [obj_type])));
                const patternInput = [exactlyOne<any>(cases_1[0]), [v]] as [any, any[]];
                return makeUnion(patternInput[0], patternInput[1]);
            }
            else {
                return oldVal;
            }
        }
        else if (typeof oldVal === "string") {
            const newStr: any = newVal;
            if (toString(newStr) === "") {
                return oldVal;
            }
            else {
                return newStr;
            }
        }
        else if (isIterable(oldVal)) {
            const innerType: any = Update_enumGetInnerType(oldVal);
            if (Update_isMapType(oldVal)) {
                return newVal;
            }
            else {
                return distinct_generic<any, any>(append_generic<any, any>(oldVal, newVal));
            }
        }
        else {
            return newVal;
        }
    }
}

export type Update_UpdateOptions_$union = 
    | Update_UpdateOptions<0>
    | Update_UpdateOptions<1>
    | Update_UpdateOptions<2>
    | Update_UpdateOptions<3>

export type Update_UpdateOptions_$cases = {
    0: ["UpdateAll", []],
    1: ["UpdateByExisting", []],
    2: ["UpdateAllAppendLists", []],
    3: ["UpdateByExistingAppendLists", []]
}

export function Update_UpdateOptions_UpdateAll() {
    return new Update_UpdateOptions<0>(0, []);
}

export function Update_UpdateOptions_UpdateByExisting() {
    return new Update_UpdateOptions<1>(1, []);
}

export function Update_UpdateOptions_UpdateAllAppendLists() {
    return new Update_UpdateOptions<2>(2, []);
}

export function Update_UpdateOptions_UpdateByExistingAppendLists() {
    return new Update_UpdateOptions<3>(3, []);
}

export class Update_UpdateOptions<Tag extends keyof Update_UpdateOptions_$cases> extends Union<Tag, Update_UpdateOptions_$cases[Tag][0]> {
    constructor(readonly tag: Tag, readonly fields: Update_UpdateOptions_$cases[Tag][1]) {
        super();
    }
    cases() {
        return ["UpdateAll", "UpdateByExisting", "UpdateAllAppendLists", "UpdateByExistingAppendLists"];
    }
}

export function Update_UpdateOptions_$reflection(): TypeInfo {
    return union_type("ISA.Aux.Update.UpdateOptions", [], Update_UpdateOptions, () => [[], [], [], []]);
}

